source/lac/petsc_parallel_block_sparse_matrix.cc

```diff
@@ -123,24 +159,65 @@ namespace PETScWrappers
     void
     BlockSparseMatrix::collect_sizes()
     {
-      BaseClass::collect_sizes();
+      auto           m = this->n_block_rows();
+      auto           n = this->n_block_cols();
+      PetscErrorCode ierr;
+
+      // Create empty matrices if needed
+      // This is neeeded by the base class
+      // not by MATNEST
+      std::vector<size_type> row_sizes(m, size_type(-1));
+      std::vector<size_type> col_sizes(n, size_type(-1));
+      std::vector<size_type> row_local_sizes(m, size_type(-1));
+      std::vector<size_type> col_local_sizes(n, size_type(-1));
+      MPI_Comm               comm = MPI_COMM_NULL;
+      for (size_type r = 0; r < m; r++)
+        {
+          for (size_type c = 0; c < n; c++)
+            {
+              if (this->sub_objects[r][c])
+                {
+                  comm = this->sub_objects[r][c]->get_mpi_communicator();
+                  row_sizes[r]       = this->sub_objects[r][c]->m();
+                  col_sizes[c]       = this->sub_objects[r][c]->n();
+                  row_local_sizes[r] = this->sub_objects[r][c]->local_size();
+                  col_local_sizes[c] =
+                    this->sub_objects[r][c]->local_domain_size();
+                }
+            }
+        }
+      for (size_type r = 0; r < m; r++)
+        {
+          for (size_type c = 0; c < n; c++)
+            {
+              if (!this->sub_objects[r][c])
+                {
+                  Mat dummy = ::create_dummy_mat(
+                    comm,
+                    static_cast<PetscInt>(row_local_sizes[r]),
+                    static_cast<PetscInt>(row_sizes[r]),
+                    static_cast<PetscInt>(col_local_sizes[c]),
+                    static_cast<PetscInt>(col_sizes[c]));
+                  this->sub_objects[r][c] = new BlockType(dummy);
```

Comment:
I get that you will get the error message somewhere in `create_dummy_mat()`, but it will say something of the form "Error 63: Invalid size -1" because at that point you have lost track of the *semantic* cause for having this invalid size. Here, however, you can give a reason:
```suggestion
      for (size_type r = 0; r < m; r++)
        {
          for (size_type c = 0; c < n; c++)
            {
              if (this->sub_objects[r][c])
                {
                  comm = this->sub_objects[r][c]->get_mpi_communicator();
                  row_sizes[r]       = this->sub_objects[r][c]->m();
                  col_sizes[c]       = this->sub_objects[r][c]->n();
                  row_local_sizes[r] = this->sub_objects[r][c]->local_size();
                  col_local_sizes[c] =
                    this->sub_objects[r][c]->local_domain_size();
                }
            }
        }
      for (size_type r = 0; r < m; r++)
        {
          for (size_type c = 0; c < n; c++)
            {
              if (!this->sub_objects[r][c])
                {
                  Assert (row_sizes[r] != size_type(-1),
                            ExcMessage ("When passing empty sub-blocks of a block matrix, you need to make "
                                                 "sure that at least one block in each block row and block column is "
                                                 "non-empty. However, block row " + std::to_string(r) + " is completely empty "
                                                 "and so it is not possible to determine how many rows it should have."));
...same for the columns...                                                 
                  Mat dummy = ::create_dummy_mat(
                    comm,
                    static_cast<PetscInt>(row_local_sizes[r]),
                    static_cast<PetscInt>(row_sizes[r]),
                    static_cast<PetscInt>(col_local_sizes[c]),
                    static_cast<PetscInt>(col_sizes[c]));
                  this->sub_objects[r][c] = new BlockType(dummy);
```
There is value in creating the exception as high up in the call stack as possible, even if you know that it would also trigger further down below.

Reply:
done